src/main/java/graphql/validation/TraversalContext.java

```diff
@@ -47,6 +48,7 @@ public class TraversalContext implements DocumentVisitor {
     private final List<GraphQLOutputType> outputTypeStack = new ArrayList<>();
     private final List<GraphQLCompositeType> parentTypeStack = new ArrayList<>();
     private final List<GraphQLInputType> inputTypeStack = new ArrayList<>();
+    private final List<InputValueWithState> defaultValueStack = new ArrayList<>();
```

Comment:
inspiration taken from graphql-js - keep a stack of default values as we traverse the AST

the older existing code was doing exactly this